pubsub/gcloud/aio/pubsub/subscriber_client.py

```diff
@@ -74,27 +75,37 @@ async def create_subscription(self,
         payload.update({'topic': topic})
         encoded = json.dumps(payload).encode()
         s = AioSession(session) if session else self.session
-        resp = await s.put(url, data=encoded, headers=headers)
+        kwargs = {
+            'timeout': timeout,
+        }
+        kwargs = {k: v for k, v in kwargs.items() if v is not None}
```

Comment:
Since we're not accepting the general `**kwargs` pattern in the function signature, this seems acceptable for adding new kwargs in the future to pass along. Is there anything else we may want to override? If not it might be simpler to just do `kwargs = {} if timeout is None else {'timeout': timeout}`

Reply:
well yeah. It just seems more welcoming for future change to add 1 key to the dicts, than changing whole expression in every function?

Reply:
Aren't `None` values acceptable for some arguments?

Reply:
@jonathan-johnston right, they are. It doesn't directly conflicts with the current code, since you can always put the key in the kwargs dict after the filter was applied, but if you think `kwargs = {} if timeout is None else {'timeout': timeout}` approach is better I'm totally ok with it

Reply:
Ah, maybe I'm misunderstanding the need for the filter on `None` values. The original code had a 10s timeout but if you allow `None` values here (which would mean `timeout=None` is piped through) then timeouts would no longer be a default 10s but infinite right? Therefore filtering out `None` values keeps the default `timeout=10`.

Reply:
Yes, `None` filter here is to ensure "not specified" semantics. But you have a point of `None` being a valid value for "timeout" param the means no timeout at all... ugh, how do we get out of wrapping hell then?

Reply:
Do we need to use `kwargs` here? Maybe just set a default value for `timeout` that matches the existing timeout of 10 seconds and pass `timeout` through directly?

Reply:
Yeah, that'll work, just feels wrong to me duplicating default value. @shaundialpad @TheKevJames opinions?

Reply:
If `None` is a valid timeout, then I suppose `kwargs` might be the best option to avoid duplicating the default value.

Reply:
I can recall @TheKevJames suggesting to avoid `kwargs` when possible. But this might actually be _that case_?

Reply:
So my general rule would be to avoid `kwargs` by default since it hides the exact arguments (ie. "explicit is better than implicit") and tends to make debugging more difficult accordingly. That's a lot more on the method arguments side of things than in calling code, mind -- ie. I appreciate that you have this method taking `session = None, timeout = None` as args instead of `**kwargs` and pulling them out from a readability perspective -- ie. this tells us that "this function cares about timeout values". That makes sense given how the `put()` method works; eg. see its method definition:
```python
@abstractmethod
async def put(self, url: str, headers: Dict[str, str], data: IO[Any],
              timeout: int) -> Response:
    pass
```
Since that function explicitly asks for a `timeout` argument, we should in turn explicitly pass it one. In that case, since we're in a different context, it makes sense to consider what our default should be here (which may just-so-happen to be a duplicate value!) and mark it as such. In other words:
```python
async def create_subscription(...,  timeout: int = 10) -> Dict[str, Any]:
    # SNIP..
    resp = await s.put(url, data=encoded, headers=headers, timeout=timeout)
```

### Why Is Duplicating This Value Ok?

So I think the whole "Don't Repeat Yourself" thing has taken on a muddled meaning over the years. "Don't Repeat Yourself" should not mean "literally do not have duplicate code anywhere". It instead should mean "do not duplicate _concepts_ which should be constant in all usecases".

_This is not one of those cases_! The session's `def put(.., timeout=10)` value is describing "what should the default timeout value be for _arbitrary put requests made to Google services_"? The subscriber's `def create_subscription(.., timeout=10)` value is describing "what should the default timeout value be for _the create subscription request in particular_"? The fact that those have identical values **is not duplication**.

### When Would We Use Kwargs?

Kwargs are super useful in cases where you want to signal that your method does not (**and should not**) care about a given argument. Most good examples along these lines have to do with inheritence-style composition: for example, consider the standard OO case of having an `Animal` class and a `Dog` subclass. You probably want the `Dog` subclass to be defined with methods along the lines of "do these  Dog-specific things... and then also whatever the more general Animal stuff is". For example:
```python
class Animal:
    def __init__(self, name: str, species: str) -> None:
        # SNIP

class Dog(Animal):
    def __init__(self, **kwargs) -> None:
        super()._init__(self, **kwargs)
```

This lets you be more explicit about which things you are changing, ie. something along the lines of:
```python
class Animal:
    def __init__(self, name: str, species: str) -> None:
        # SNIP

class Dog(Animal):
    def __init__(self, has_dog_breath: bool, **kwargs) -> None:
        self._needs_a_toothbrush = has_dog_breath
        super()._init__(self, **kwargs)
```

In this specific example, using `kwargs` in the `create_subscription()` argument would be tying the implementation details together too closely -- it would be akin to saying "whatever extra stuff the put method does"... which would be pretty confusing and implicit to users, since they then need to know that we just-so-happen to be using `Session.put` later on in the method, which is an implementation detail our users should not have to care about!

Reply:
@TheKevJames That's ok, I don't mind duplicating this value at this point.

but for the joy of intellectual discussion let me clarify one point! I don't like having a stack of wrappers not because I think it violates DRY (anyone who saw my tests knows I have no problems hitting cmd+v hitting cmd+v) but mostly because it _feels wrong_. And I think this time there is an actual reason to this.

>it would be akin to saying "whatever extra stuff the put method does"... which would be pretty confusing and implicit to users, since they then need to know that we just-so-happen to be using Session.put later on in the method, which is an implementation detail our users should not have to care about!

I think we already have a much more confusing behavior. If you need to browse the source code to understand what's hiding behind the `kwargs`, if you have to know that the package uses `aiohttp` under the hood to effectively work with it – I agree this is a bit inconvenient, but ultimately this is a virtue because in expands user's true knowledge about the system.
What we did is that we restricted api flexibility by being overly zealous about explicit args. `aiohttp` has a ton of parameters one can customize per-request, but we decided we know better. There are actually 3 timeout parameters, but we only allow one of them.  What I find the most confusing is `session` parameter. Our type hints clearly suggests that we expect an optional `aiohttp.ClientSession` session. A user can put a lot of per-session parameters in there and most of them will work... but not `timeout`, because we force override it on per-request level. And this is truly confusing because a user will rightly assume that session-level timeout will work, but it won't because of an implementation detail.

tl;dr I think saying "this is our method and by the way we will happily pass any kwargs to `aiohttp`, so please consult its docs" is better then "this is out method's signature and that's all you need to know"

Reply:
That's a great point, Alex! I do agree that we could benefit from expanding this (although, perhaps with some renames; eg. something like `**http_kwargs` might make more sense, since our code is not 100% strictly passthroughs only to http requests and IIRC we have cases were the kwargs are used for other things).

I think the main point of contention for me is that this is not how the library _currently_ works. I definitely agree that allowing users this flexible would be a benefit, but I think keeping things consistent is more valuable than having this only be the case for _some_ methods; I'd support switching the `gcloud-aio` libs over to this style more holistically, but not here and there as we modify things, since having each method behave differently would be quite confusing!

As to whether the value of exposing these params is worth the sum effort of changing everything? _Probably_, yeah, though for it to fit into our scheduled non-20% work we'd have to be running into places where we'd make use of those params ourselves :P

Reply:
totally agree with `http_kwargs`. And yeah, not worth to change everything at this point, those are mostly consideration for the next time we will design client libraries or anything like that.